Add support to Interface based model

handreyrc opened the pull request:
Fixes #194

This PR intoduces interfaces to model definitions allowing to work with plain objects instead of class instances.
This approach makes it easier handling changes in the model and make it compatible with libs such as immer.

The idea behind this implementation is to keep classes under the hood to handle model properties, however, only plain objects are exposed by unmarshalling JSON/YAML or by using the SKD builders. To achieve that, lodash lib was used to convert class instances to plain shallow copies of the original class instances and so preserving properties and methods. 

No changes in the API were be introduced by this PR.
Old and deprecated dependencies of the project were updated.
Immer dependency is only being consumed by in the unit tests.

handreyrc committed:
Add interfaces and plain objects based model

Signed-off-by: handreyrc <handrey.cunha@gmail.com>

ricardozanini reviewed (commented):
My only concern is with the changes you made to the schema. Why were they necessary? The rest seems pretty straightforward to me.

handreyrc commented:
@ricardozanini,

The changes to the schema were made automatically by the script "update-code-base". That was the first thing I did before start working on the project.
It donwloads the schemas, then it recreates definitions (this is an intermediate step, the actual definitions exposed by the sdk are not auto-generated) and builders based on the downloded schemas.

Please, @ricardozanini @JBBianchi, let me know if it is an issue. 
If so, I can port my changes on top of the orignal code and run the script recreate the builders only. It's not a big deal.

Thanks a lot!

handreyrc commented:
@ricardozanini,
Just adding more information about the schema. The script I mentioned above is downloading the schema from this URL:
https://raw.githubusercontent.com/serverlessworkflow/specification/0.8.x/schema/workflow.json

ricardozanini commented:
@JBBianchi is on PTO until next week; we will hold this PR until then.

JBBianchi commented on tests/lib/workflow-validator.spec.ts:
```diff
@@ -44,6 +44,8 @@ describe('workflow-validator, invalid state', () => {
     workflowWithEmptyStates.states = [];
 
+    console.log('--> handrey: ' + workflowWithEmptyStates.states.length);
+
```
Debugging artifact ?

```suggestion
```

handreyrc replied:
@JBBianchi,
Sorry, I bypassed this one. It's removed.

JBBianchi commented on package.json:
```diff
@@ -39,34 +39,36 @@
+    "immer": "^10.0.3",
+    "jest": "^30.2.0",
     "lint-staged": "11.0.0",
+    "lodash": "^4.17.21",
```
Use [lodash.toplainobject](https://www.npmjs.com/package/lodash.toplainobject) instead?

handreyrc replied:
@JBBianchi,
I agree! It's done!

JBBianchi reviewed (changes requested):
I’m fine with this approach. Just a small note: since normalize() now returns a plain object (e.g. IWorkflow) instead of an instance, runtime checks like `instanceof Workflow` will no longer work on the returned value. That’s totally expected for this use case, but it’s worth keeping in mind for any downstream logic that might rely on class instances.

handreyrc committed:
fix lodash dependency and remove debug message

Signed-off-by: handreyrc <handrey.cunha@gmail.com>

handreyrc committed:
do not break the SDK for those relying on classes

Signed-off-by: handreyrc <handrey.cunha@gmail.com>

handreyrc commented:
> I’m fine with this approach. Just a small note: since normalize() now returns a plain object (e.g. IWorkflow) instead of an instance, runtime checks like `instanceof Workflow` will no longer work on the returned value. That’s totally expected for this use case, but it’s worth keeping in mind for any downstream logic that might rely on class instances.

@JBBianchi, 
Thanks a lot for your guidance here. I tried to not add anything new to the SDK, however, if we want to not cause any issues to those relying on classes it is not possible. I changed my approach keeping in mind that those relying on classes shall not be affected at all.  
* I added a new method ''asPlainObject()" to definition classes so no changes are necessary for those relying on classes.  For those working with interfaces just call this same method after the build() and it will do the trick.
* I also added an optional parameter to "fromSource" method in the workflow definition class, so it will work with classes by default, however, by setting the optional parameter "asPlainObject" to true cause the workflow to be deserialized to a plain object. 
* One last change, the "normalize" method now works with classes and plain objects. Basically, the original object instance is evaluated to be a class or a plain object and then it will behave accordingly to input type.

That being said, I think we are good for another round.

Thanks reviewing this work!

JBBianchi commented:
Great job @handreyrc, thanks for addressing my previous remark.

Could you please also update the package version (and possibly its name) so we can publish it?
- Either, if you want to keep the current name (even though it's incorrect), please bump the version to `3.0.4` 
- Or rename to `@serverlessworkflow/sdk` and align the version with the spec & release: `0.8.4`.

handreyrc committed:
fix wrong name and version

Signed-off-by: handreyrc <handrey.cunha@gmail.com>

handreyrc commented:
> Great job @handreyrc, thanks for addressing my previous remark.
> 
> Could you please also update the package version (and possibly its name) so we can publish it?
> 
> * Either, if you want to keep the current name (even though it's incorrect), please bump the version to `3.0.4`
> * Or rename to `@serverlessworkflow/sdk` and align the version with the spec & release: `0.8.4`.

@JBBianchi,  I'm always for fixing what is wrong. It's a good opportunity to make it right so lets go for it.
* Changed name from @severlessworkflow/sdk-typescript to @serverlessworkflow/sdk-typescript
* Fixed wrong name references in the README file
* Changed version from 3.0.3 to 0.8.4 

Please let me know if anything else is missing and I promptly adjust it.
Thanks!
